dhis-2/dhis-services/dhis-service-expression-parser/src/main/java/org/hisp/dhis/expressionparser/ExpressionVisitor.java

```diff
@@ -0,0 +1,596 @@
+ * expression: With the "listener" pattern, ANTRL traverses the parsed tree
+ * but allows a subclass to listen at each node as it is processed. With the
+ * "visitor" pattern, the subclass (such as this class) is called for each node,
+ * and it is up to the code in this pattern to make explicit calls to descendant
+ * nodes. To process DHIS2 expressions the visitor pattern is used, because the
+ * visiting of descendant nodes can be used to advantage as described here.
+ * <p/>
+ * A left-recursive descent parser such as ANTLR is designed to evaluate
+ * expressions left-to-right, but the nodes are actually visited initially in
+ * the order right-to-left. For example, if the expression is "1 + 2 - 3", the
+ * higher-level node "(subexpression) - 3", is visited first, then it would
+ * visit the lower-level subexpression node "1 + 2" to evaluate that
+ * subexpression. After the subexpression returns its value, to the higher-level
+ * node, it finally subtracts 3 from that returned subexpression value.
+ * <p/>
+ * As an example of how this class makes use of the right-to-left visiting,
+ * consider the expression #{uid}.period(-3,-1).sum(). The sum() function is
+ * visited first as (subexpression).sum(). The sum() function will expect its
+ * subexpression to return multiple values to be summed, and so evaluates its
+ * subexpresssion. In doing so, the period() function is next visited as
+ * (subexpression).period(-3,-1). The period function then iterates over periods
+ * -3 to -1, and calls multiple times to evaluate its (subexpression) for each
+ * of these periods in turn. Finally, it returns an object with the multiple
+ * values from these evaluations to the sum() function, which then sums them.
+ * <p/>
+ * If the ANTLR listener pattern were used in the above example, the
+ * subexpression to the left of the period() function would be called only
+ * once, as determined by ANTLR. But by using the visitor period, this class
+ * is able to do things like evaluate the subexpression multiple times.
+ * <p/>
+ * This is an abstract class, which is intended to be extended for two different
+ * purposes: (1) an "items visitor" which collects the expression items as
+ * visited before data is fetched from the database, to know which values need
+ * to be fetched, and which periods and organisaiton units they need fetching
+ * for, and (2) an "value visitor" once the database values have been found, to
+ * plug them into the expression and compute the total value of the expression.
+ *
+ * @author Jim Grace
+ */
+public abstract class ExpressionVisitor
+    extends ExpressionBaseVisitor<Object>
+{
+    @Autowired
+    protected IdentifiableObjectManager manager;
+
+    @Autowired
+    protected ConstantService constantService;
+
+    protected Map<String, Double> constantMap = null;
+
+    protected Map<String, Integer> orgUnitCountMap = null;
+
+    protected Double days = DUMMY_VALUE;
```

Comment:
This was not used as the default number of days in computing an expression value. It was only used as a placeholder value when evaluating an expression syntax in ExpressionItemsVisitor, and is set to the same dummy value as other unknowns such as the values of DimensionalItemObjects. The default number of days used for computing an expression value is zero, for compatibility with the current DefaultExpressionService code. This had been set in ExpressionValueVisitor line 74 as DUMMY_DAYS, where this comment is well taken; it should have been called DEFAULT_DAYS.

I have now simplified this. Initially I was setting days to a dummy value of 1 to avoid possible problems with division by zero during ExpressionItemsVisitor. But this was unnecessary since dividing by 0 gives a valid Double value (infinity) and so can still be used for type-checking. I have simplified this now by setting days to DEFAULT_DAYS in the ExpressionVisitor abstract class, and not overwriting it in ExpressionValueVisitor with a different default if there is no period.

DUMMY_VALUE could be renamed to PLACEHOLDER_VALUE, or PARSING_PLACEHOLDER_VALUE, or some such, if we want to avoid the word DUMMY.